### PR TITLE
322 add module deseq2

### LINKED ISSUE #322: add module DESEQ2
- [ ] add module DESEQ2 into the experimental branch

### PR DESCRIPTION
**Pull request by roskamsh**
## PR checklist

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/differentialabundance/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, also make a PR on the nf-core/differentialabundance _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [ ] Make sure your code lints (`nf-core lint`).
- [ ] Ensure the test suite passes (`nf-test test main.nf.test -profile test,docker`).
- [ ] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [ ] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: subworkflows/local/differential/main.nf [hunk 1]
```diff
@@ -73,24 +75,70 @@ workflow DIFFERENTIAL {
...
+        .join(ch_samplesheet)
+        .combine(ch_contrasts)
+        .combine(ch_transcript_lengths)
+        .combine(ch_control_features)
+        .combine(ch_tools_single.deseq2)
+        .multiMap {
+            meta_data, counts, samplesheet, meta_contrast, contrast_variable, reference, target, meta_lengths, lengths, meta_control, control, pathway, meta_tools ->
+                def meta = meta_data.clone() + meta_contrast.clone() + meta_lengths.clone() + meta_control.clone() + meta_tools.clone()
+                contrast: [ meta, contrast_variable, reference, target ]
+                samplesheet: [ meta, samplesheet, counts ]
```
Comment (WackerO) [inline subworkflows/local/differential/main.nf]:
```suggestion
                samples_and_matrix: [ meta, samplesheet, counts ]
```

File: subworkflows/local/differential/main.nf [hunk 2]
```diff
@@ -73,24 +75,70 @@ workflow DIFFERENTIAL {
...
+                control_features:   [ meta, control ]
+                transcript_lengths: [ meta, lengths ]
+                pathway: [ meta, pathway ]
+        }
+        .set { ch_deseq2 }
+
+    // do we need this process DESEQ2_NORM?
+    DESEQ2_NORM (
+            ch_deseq2.contrast.first(),
+            ch_deseq2.samplesheet,
```
Comment (WackerO) [inline subworkflows/local/differential/main.nf]:
```suggestion
            ch_deseq2.samples_and_matrix,
```

File: subworkflows/local/differential/main.nf [hunk 3]
```diff
@@ -73,24 +75,70 @@ workflow DIFFERENTIAL {
...
+            ch_deseq2.control_features,
+            ch_deseq2.transcript_lengths
+        )
+
+    DESEQ2_DIFFERENTIAL (
+            ch_deseq2.contrast,
+            ch_deseq2.samplesheet,
```
Comment (WackerO) [inline subworkflows/local/differential/main.nf]:
```suggestion
            ch_deseq2.samples_and_matrix,
```

File: subworkflows/local/differential/main.nf [hunk 4]
```diff
@@ -73,24 +75,70 @@ workflow DIFFERENTIAL {
...
+                def meta = meta_data.clone() + meta_contrast.clone() + meta_lengths.clone() + meta_control.clone() + meta_tools.clone()
+                contrast: [ meta, contrast_variable, reference, target ]
+                samplesheet: [ meta, samplesheet, counts ]
+                control_features:   [ meta, control ]
+                transcript_lengths: [ meta, lengths ]
+                pathway: [ meta, pathway ]
+        }
+        .set { ch_deseq2 }
+
+    // do we need this process DESEQ2_NORM?
```
Comment (WackerO) [inline subworkflows/local/differential/main.nf:98]: Yes, we do 🙂

**Comment by WackerO** [comment]
Hey @roskamsh, thanks for taking over this issue! Nice work, just a couple of small changes.

@suzannejin I just had a quick look at the output and was wondering if you know why the DESeq2 result tables differ slightly in the `test_experimental` profile when compared to the `test` profile:

test:
```
gene_id	baseMean	log2FoldChange	lfcSE	pvalue	padj
ENSMUSG00000023978	309.50335903	5.015358	2.7289078	NA	NA
ENSMUSG00000030324	1881.77922741	0.9717132	2.0668858	NA	NA
ENSMUSG00000031450	195.24652216	0.8356462	1.90527993	NA	NA
```

experimental:
```
gene_id	baseMean	log2FoldChange	lfcSE	pvalue	padj
ENSMUSG00000023978	308.89789121	5.406014	2.57197697	NA	NA
ENSMUSG00000030324	1763.85049182	1.057522	2.14925341	NA	NA
ENSMUSG00000031450	201.02654536	1.131211	2.19318651	NA	NA
```

**Comment by roskamsh** [comment]
> Hey @roskamsh, thanks for taking over this issue! Nice work, just a couple of small changes.
> 
> @suzannejin I just had a quick look at the output and was wondering if you know why the DESeq2 result tables differ slightly in the `test_experimental` profile when compared to the `test` profile:
> 
> test:
> 
> ```
> gene_id	baseMean	log2FoldChange	lfcSE	pvalue	padj
> ENSMUSG00000023978	309.50335903	5.015358	2.7289078	NA	NA
> ENSMUSG00000030324	1881.77922741	0.9717132	2.0668858	NA	NA
> ENSMUSG00000031450	195.24652216	0.8356462	1.90527993	NA	NA
> ```
> 
> experimental:
> 
> ```
> gene_id	baseMean	log2FoldChange	lfcSE	pvalue	padj
> ENSMUSG00000023978	308.89789121	5.406014	2.57197697	NA	NA
> ENSMUSG00000030324	1763.85049182	1.057522	2.14925341	NA	NA
> ENSMUSG00000031450	201.02654536	1.131211	2.19318651	NA	NA
> ```

Could this be due to the new parameter I added in the test_experimental.config file? When the DESEQ2 process ran as initially specified, this line (https://github.com/nf-core/differentialabundance/blob/dev-ratio/modules/nf-core/deseq2/differential/templates/deseq_de.R#L486) threw an error, saying that nsub was greater that nrow. I went in and replicated the issue myself, and it seems that the input matrix only contained 982 features, whereas the `vst_nsub` default value was set to 1000. In order to resolve this I had to change the nextflow parameter `vst_nsub` in the test_experimental.config file. I am not really sure why this wasn't found in the `test` run itself, if the input data is exactly the same?

Is the input data exactly the same? Even so, this would only really change the `*.vst.tsv` output and not the `*.deseq2.results.tsv` output.

**Comment by WackerO** [comment]
> > Hey @roskamsh, thanks for taking over this issue! Nice work, just a couple of small changes.
> > @suzannejin I just had a quick look at the output and was wondering if you know why the DESeq2 result tables differ slightly in the `test_experimental` profile when compared to the `test` profile:
> > test:
> > ```
> > gene_id	baseMean	log2FoldChange	lfcSE	pvalue	padj
> > ENSMUSG00000023978	309.50335903	5.015358	2.7289078	NA	NA
> > ENSMUSG00000030324	1881.77922741	0.9717132	2.0668858	NA	NA
> > ENSMUSG00000031450	195.24652216	0.8356462	1.90527993	NA	NA
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > experimental:
> > ```
> > gene_id	baseMean	log2FoldChange	lfcSE	pvalue	padj
> > ENSMUSG00000023978	308.89789121	5.406014	2.57197697	NA	NA
> > ENSMUSG00000030324	1763.85049182	1.057522	2.14925341	NA	NA
> > ENSMUSG00000031450	201.02654536	1.131211	2.19318651	NA	NA
> > ```
> 
> Could this be due to the new parameter I added in the test_experimental.config file? When the DESEQ2 process ran as initially specified, this line (https://github.com/nf-core/differentialabundance/blob/dev-ratio/modules/nf-core/deseq2/differential/templates/deseq_de.R#L486) threw an error, saying that nsub was greater that nrow. I went in and replicated the issue myself, and it seems that the input matrix only contained 982 features, whereas the `vst_nsub` default value was set to 1000. In order to resolve this I had to change the nextflow parameter `vst_nsub` in the test_experimental.config file. I am not really sure why this wasn't found in the `test` run itself, if the input data is exactly the same?
> 
> Is the input data exactly the same? Even so, this would only really change the `*.vst.tsv` output and not the `*.deseq2.results.tsv` output.

I thought so too but tested the two profiles with the same nsub values and the difference was still there. But I just realized that in the experimental config, no `--transcript_length_matrix` is provided. I tested it, that is indeed the source of the differences, so all is good :)

Thanks for the work!

**Review by WackerO — APPROVED** [review]

File: subworkflows/local/differential/main.nf (hunk 4 above)
Comment (suzannejin) [reply subworkflows/local/differential/main.nf:98]: why do we need both deseq2_norm and deseq2_differential? Isn't deseq2_differential also outputing the normalized data?

**Comment by suzannejin** [comment]
@roskamsh thank you so much for the contribution! 
And thank you @WackerO for reviewing!
Lgtm too! I will merge now.

**Pull request merged**